tensorflow/core/kernels/matmul_op_test.cc

```diff
@@ -214,7 +224,8 @@ class FusedMatMulOpTest : public OpsTestBase {
     ASSERT_EQ(matmul.dtype(), fused_matmul.dtype());
     ASSERT_EQ(matmul.shape(), fused_matmul.shape());
 
-    test::ExpectClose(matmul, fused_matmul, /*atol=*/1e-5);
+    double atol = this->TValueType == DT_HALF ? 1e-1 : 1e-5;
```

Comment:
Would it be possible to also pass a rtol value instead of increasing the atol value to 1e-1 (which seems rather big)?

Reply:
yes, the error seems to be rather big for Eigen::half datatype, especially for the test Matmul256x128x64.
Then I need to investigate this further, so far I can only reduce the 'atol' to 1e-2.
```
i = 0 Tx[i] = 2.0625 Ty[i] = 2.060546875
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (4.296875 not close to 4.30078125)
Expected: true
i = 85 Tx[i] = 4.296875 Ty[i] = 4.30078125
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (4.359375 not close to 4.36328125)
Expected: true
i = 96 Tx[i] = 4.359375 Ty[i] = 4.36328125
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (4.421875 not close to 4.42578125)
Expected: true
i = 181 Tx[i] = 4.421875 Ty[i] = 4.42578125
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (2.12109375 not close to 2.119140625)
Expected: true
i = 187 Tx[i] = 2.12109375 Ty[i] = 2.119140625
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (2.078125 not close to 2.076171875)
Expected: true
i = 229 Tx[i] = 2.078125 Ty[i] = 2.076171875
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (4.5078125 not close to 4.50390625)
Expected: true
i = 243 Tx[i] = 4.5078125 Ty[i] = 4.50390625
```
I have also played around this 'rtol' values but could not figure out a meanigful value which works fine for all cases

Reply:
So I guess a combination of higher atol + specifying a similar rtol value should likely give the best result.
Can you try for example atol=1e-3 rtol=1e-3?

Reply:
yes, surpringly, with these values of atol/rtol as you suggested, the test for Eigen::half also works fine (though it takes ages because CUDNN does some autotuning first to find the proper algorithm to use):
```
[----------] 8 tests from Test/FusedMatMulWithBiasOpTest/1, where TypeParam = Eigen::half
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x128x64
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x128x64 (149 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x256
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x256 (131 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x256x1
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x256x1 (89 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x1
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x1 (19 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x128x64WithActivation
2023-10-05 10:14:04.231568: I external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:440] Loaded cuDNN version 8904
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x128x64WithActivation (220203 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x256WithActivation
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x256WithActivation (46056 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x256x1WithActivation
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x256x1WithActivation (0 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x1WithActivation
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x1WithActivation (0 ms)
[----------] 8 tests from Test/FusedMatMulWithBiasOpTest/1 (266651 ms total)

[----------] Global test environment tear-down
[==========] 16 tests from 2 test suites ran. (269672 ms total)
[  PASSED  ] 16 tests.
```